pkg/testing/test_runner.go

```diff
@@ -0,0 +1,281 @@
+package testing
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path"
+	"time"
+
+	"github.com/symbiosis-cloud/cli/pkg/identity"
+	"github.com/symbiosis-cloud/cli/pkg/output"
+	"github.com/symbiosis-cloud/cli/pkg/symcommand"
+	"golang.org/x/sync/errgroup"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubernetes "k8s.io/client-go/kubernetes"
+)
+
+type TestState string
+
+const (
+	TEST_STATE_FAILED  TestState = "FAILED"
+	TEST_STATE_SUCCESS TestState = "SUCCESS"
+	TEST_STATE_PENDING TestState = "PENDING"
+)
+
+type TestJob struct {
+	Image    string    `json:"image"`
+	Commands []string  `json:"commands"`
+	State    TestState `json:"state"`
+}
+
+type TestResult struct {
+	Name     string        `json:"name"`
+	Image    string        `json:"image"`
+	Commands []string      `json:"commands"`
+	State    TestState     `json:"state"`
+	ExitCode int32         `json:"exitCode"`
+	Logs     string        `json:"logs"`
+	Duration time.Duration `json:"duration"`
+}
+
+type TestRunner struct {
+	jobs        []*TestJob
+	identity    *identity.ClusterIdentity
+	clientSet   *kubernetes.Clientset
+	CommandOpts *symcommand.CommandOpts
+}
+
+func (t *TestRunner) Run(testOutputDir string) error {
+
+	t.CommandOpts.Logger.Info().Msgf("Running %d tests...", len(t.jobs))
+
+	// TODO: make timeouts configurable
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Minute*30)
+
+	results := make([]*TestResult, len(t.jobs))
+
+	defer cancel()
+
+	errGroup := new(errgroup.Group)
+	podsApi := t.clientSet.CoreV1().Pods(t.CommandOpts.Namespace)
+
+	var deletePods []string
+
+	for i, job := range t.jobs {
+
+		podName := fmt.Sprintf("test-job-%d", i)
+		deletePods = append(deletePods, podName)
+		results[i] = &TestResult{
+			Name:     fmt.Sprintf("test-%d", i),
+			Image:    job.Image,
+			Commands: job.Commands,
+			State:    TEST_STATE_PENDING,
+		}
+
+		err := results[i].Write(testOutputDir)
+
+		if err != nil {
+			return err
+		}
+
+		podSpec := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: t.CommandOpts.Namespace,
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:    fmt.Sprintf("test-container-%d", i),
+						Image:   job.Image,
+						Command: job.Commands,
+					},
+				},
+				RestartPolicy: v1.RestartPolicyNever,
+			},
+		}
+
+		num := i
+		executionStart := time.Now()
+
+		errGroup.Go(func() error {
```

Comment:
This is a veeery long function, maybe it can be broken up a little bit?

Reply:
@Styren I split the method into two, can you check ?

Reply:
Much better! 🙌